wordpress-mobile/WordPress-Android
# Reader: Show preview images in lists for more posts

Fixes #12894, allowing the reader to find suitable preview images for posts more often.

#### Feed list

![reader-android-feed-beforeafter](https://user-images.githubusercontent.com/9613966/92977369-cb051a80-f4c7-11ea-837e-b2afeaec2ce7.png)

#### Site list

![reader-android-site-beforeafter](https://user-images.githubusercontent.com/9613966/92296916-c9da6780-ef74-11ea-8987-edc345e7e4b3.png)

This PR has two parts:

### 1. Fixing an accidental regex breakage

A [huge checkstyle refactor from two years ago](https://github.com/wordpress-mobile/WordPress-Android/commit/ed574ea833797f71e28fb2abd060fee7d305376c#diff-981ccd25a498c16f8b2887442812b346) ended up accidentally adding spaces to a few regex patterns used by the Reader to parse images. This almost entirely broke support for preview images that weren't set as featured. I undid the breakage in 0379e37 (and added some tests).

### 2. Adding support for `srcset` and `data-large-file` `img` tag attributes

The above fix improves things a bit but still fails to recognize images in several kinds of posts. This is especially noticeable in the site view. As explained in #12894 the `/read/feed/` and `/read/sites/` endpoints return different content sometimes. For some reason one of the differences is that some images have no width attribute in the `/read/sites/` endpoint, so what works in the feed doesn't work sometimes in the site view.

To mitigate this, I added support for a few new `img` tag attributes that seem pretty common in WP posts: [srcset](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images) and data-large-file. These are checked for in a fallback style.

### To test:

With and without this patch, follow a few sites with a variety of posts containing images and compare.

Mainly looking for posts with a story block, and posts with image blocks (but no featured image set). (I can recommend a few.)

(For Automatticians: please proxy on your device to fully test posts with the Story block, since depending on the endpoint those won't work correctly otherwise. See pbArwn-14B-p2.)

You can also check the regex fix individually by checking out 86dedbe first and running the (failing) unit tests added.

Notes:
* Some stories (where the first slide is a video) still won't show a preview image - I'll look into that separately
* Detail view of stories is broken right now

### PR submission checklist:
